li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaFederatedProducerImpl.java

```diff
@@ -355,7 +185,35 @@ public void close() {
 
   @Override
   public void close(long timeout, TimeUnit timeUnit) {
-    throw new UnsupportedOperationException("Not implemented yet");
+    if (_producers.isEmpty()) {
+      LOG.warn("No producers to close for cluster group {}", _clusterGroup);
+      return;
+    }
+
+    LOG.info("Closing LiKafkaProducer for cluster group {} in {} {}...", _clusterGroup, timeout, timeUnit);
+
+    long startTimeMs = System.currentTimeMillis();
+    long deadlineTimeMs = startTimeMs + timeUnit.toMillis(timeout);
```

Comment:
Possible overflow: when `close()` is invoked, the timeout value is `Integer.MAX_VALUE`. Is it safe to compute deadline as `startTimeMs + timeUnit.toMillis(timeout)` ? Fwiw, I understand this is similar in the existing `LiKafkaProducerImpl`.

Reply:
Given that the timeout is Integer.MAX_VALUE and we are using long for the deadline, I think it should be safe (at least as long as we live.. :P). I admit that this is not the most elegant way, though. I also noticed the open source Kafka code also does this.